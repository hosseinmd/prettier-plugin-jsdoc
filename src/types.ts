import { ParserOptions } from "prettier";

export interface JsdocOptions {
  jsdocSpaces: number;
  jsdocPrintWidth?: number;
  jsdocDescriptionWithDot: boolean;
  jsdocDescriptionTag: boolean;
  jsdocVerticalAlignment: boolean;
  jsdocKeepUnParseAbleExampleIndent: boolean;
  /** default is true */
  jsdocSingleLineComment: boolean;
  jsdocSeparateReturnsFromParam: boolean;
  jsdocSeparateTagGroups: boolean;
  jsdocAddDefaultToDescription: boolean;
  jsdocCapitalizeDescription: boolean;
  tsdoc: boolean;
}

export interface AllOptions extends ParserOptions, JsdocOptions {}

type LocationDetails = { line: number; column: number };
type Location = { start: LocationDetails; end: LocationDetails };

export type PrettierComment = {
  type: "CommentBlock" | "Block";
  value: string;
  start: number;
  end: number;
  loc: Location;
};

export type Token = {
  type:
    | "CommentBlock"
    | "Block"
    | {
        label: string; // "function" | "name";
        keyword?: string;
        beforeExpr: boolean;
        startsExpr: boolean;
        rightAssociative: boolean;
        isLoop: boolean;
        isAssign: boolean;
        prefix: boolean;
        postfix: boolean;
        binop: null;
      };
  value: string;
  start: number;
  end: number;
  loc: Location;
};

export type AST = {
  start: number;
  end: number;
  loc: Location;
  errors: [];
  program: {
    type: "Program";
    start: number;
    end: number;
    loc: [];
    sourceType: "module";
    interpreter: null;
    body: [];
    directives: [];
  };
  comments: PrettierComment[];
  tokens: Token[];
};
