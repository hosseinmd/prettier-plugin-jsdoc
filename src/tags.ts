const ABSTRACT = "abstract";
const ASYNC = "async";
const AUGMENTS = "augments";
const AUTHOR = "author";
const CALLBACK = "callback";
const CATEGORY = "category";
const CLASS = "class";
const CONSTANT = "constant";
const DEFAULT = "default";
const DEPRECATED = "deprecated";
const DESCRIPTION = "description";
const EXAMPLE = "example";
const EXTENDS = "extends";
const EXTERNAL = "external";
const FILE = "file";
const FIRES = "fires";
const FLOW = "flow";
const FUNCTION = "function";
const MEMBER = "member";
const MEMBEROF = "memberof";
const OVERRIDE = "override";
const PARAM = "param";
const PRIVATE = "private";
const PROPERTY = "property";
const PROVIDES_MODULE = "providesModule";
const RETURNS = "returns";
const SEE = "see";
const SINCE = "since";
const TEMPLATE = "template";
const THROWS = "throws";
const TODO = "todo";
const TYPE = "type";
const TYPEDEF = "typedef";
const VERSION = "version";
const YIELDS = "yields";

export {
  ABSTRACT,
  ASYNC,
  AUGMENTS,
  AUTHOR,
  CALLBACK,
  CATEGORY,
  CLASS,
  CONSTANT,
  DEFAULT,
  DEPRECATED,
  DESCRIPTION,
  EXAMPLE,
  EXTENDS,
  EXTERNAL,
  FILE,
  FIRES,
  FLOW,
  FUNCTION,
  MEMBER,
  MEMBEROF,
  OVERRIDE,
  PARAM,
  PRIVATE,
  PROPERTY,
  PROVIDES_MODULE,
  RETURNS,
  SEE,
  SINCE,
  TEMPLATE,
  THROWS,
  TODO,
  TYPE,
  TYPEDEF,
  VERSION,
  YIELDS,
};
