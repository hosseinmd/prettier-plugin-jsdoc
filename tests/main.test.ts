import {format} from "prettier";
import { AllOptions } from "../src/types";

function subject(code: string, options: Partial<AllOptions> = {}) {
  return format(code, {
    parser: "babel",
    plugins: ["prettier-plugin-jsdoc"],
    jsdocSpaces: 1,
    ...options,
  } as AllOptions);
}

test("JS code should be formatted as usuall", async () => {
  const result = await subject(`
const variable1 = 1             // No semicolon
const stringVar = "text"        // Wrong quotes
  const indented = 2            // Wrong indentation

// Longer then 80 characters
const someLongList = ['private', 'memberof', 'description', 'example', 'param', 'returns', 'link']`);

  expect(result).toMatchSnapshot();
});

test("Should format regular jsDoc", async () => {
  const result = await subject(`
/**
* function example description that was wrapped by hand
* so it have more then one line and don't end with a dot
* REPEATED TWO TIMES BECAUSE IT WAS EASIER to copy
* function example description that was wrapped by hand
* so it have more then one line.
* @return {Boolean} Description for @returns with s
* @param {String|Number} text - some text description that is very long and needs to be wrapped
* @param {String} [defaultValue="defaultTest"] TODO
* @arg {Number|Null} [optionalNumber]
* @private
*@memberof test
@async
* @examples
*   var one = 5
*   var two = 10
*
*   if(one > 2) { two += one }
* @undefiendTag${" "}
* @undefiendTag {number} name des
*/
const testFunction = (text, defaultValue, optionalNumber) => true
`);

  expect(result).toMatchSnapshot();
  expect(await subject(result)).toMatchSnapshot();
});

test("Should format jsDoc default values", async () => {
  const result = await subject(`
/**
* @param {String} [arg1="defaultTest"] foo
* @param {number} [arg2=123] the width of the rectangle
* @param {number} [arg3= 123 ]
* @param {number} [arg4= Foo.bar.baz ]
* @param {number|string} [arg5=123] Something. Default is \`"wrong"\`
*/
`);

  expect(result).toMatchSnapshot();
  expect(await subject(result)).toMatchSnapshot();

  const result2 = await subject(
    `
/**
* @param {String} [arg1="defaultTest"] foo
* @param {number} [arg2=123] the width of the rectangle
* @param {number} [arg3= 123 ]
* @param {number} [arg4= Foo.bar.baz ]
* @param {number|string} [arg5=123] Something. Default is \`"wrong"\`
*/
`,
    {
      jsdocAddDefaultToDescription: false,
    },
  );

  expect(result2).toMatchSnapshot();
});

test("Should convert to single line if necessary", async () => {
  const Result1 = await subject(`/** single line description*/`);
  const Result2 = await subject(`/**
 * single line description
 * @example
 */`);

  const Result3 = await subject(`/**
 * single line description
 * @return {Boolean} Always true
 * @example
 */`);

  expect(Result1).toMatchSnapshot();
  expect(Result2).toMatchSnapshot();
  expect(Result3).toMatchSnapshot();
});

test("Should convert to single multiLine", async () => {
  const Result1 = await subject(`/** single line description*/`, {
    jsdocSingleLineComment: false,
  });
  const Result2 = await subject(
    await subject(`/**
 * single line description
 * @example
 */`),
    {
      jsdocSingleLineComment: false,
    },
  );

  const Result3 = await subject(
    `/**
 * single line description
 * @return {Boolean} Always true
 * @example
 */`,
    {
      jsdocSingleLineComment: false,
    },
  );

  expect(Result1).toMatchSnapshot();
  expect(Result2).toMatchSnapshot();
  expect(Result3).toMatchSnapshot();
});

test(" undefined|null|void type", async () => {
  const Result1 = await subject(`/**
 * @return {undefined}
 */`);

  const Result2 = await subject(`/**
 * @return {null}
 */`);

  const Result3 = await subject(`/**
 * @returns { void }${" "}
 */`);

  expect(Result1).toMatchSnapshot();
  expect(Result2).toMatchSnapshot();
  expect(Result3).toMatchSnapshot();
});

test("Should keep defined inner types", async () => {
  const Result1 = await subject(`/**
 * @param {Array.<String>} test test param
 */`);

  const Result2 = await subject(`/**
 * @param {String[]} test Test param
 */`);

  const Result3 = await subject(`/**
 * @param {(String|Object)[]} test Test param
 */`);

  const Result4 = await subject(`/**
 * @returns {Promise<Number|String|undefined>} test promise
 */`);

  const Result5 = await subject(`/**
 * @returns {Object<Number|String|undefined>} test object
 */`);

  expect(Result1).toMatchSnapshot();
  expect(Result2).toMatchSnapshot();
  expect(Result3).toMatchSnapshot();
  expect(Result4).toMatchSnapshot();
  expect(Result5).toMatchSnapshot();
});

test("Sould keep params ordering when more than 10 tags are present", async () => {
  const Result1 = await subject(`/**
 * description
 * @param {Number} test1 Test param
 * @param {Number} test2 Test param
 * @param {Number|String} test3 Test param
 * @param {?undefined} test4 Test param
 * @param {!undefined} test5 Test param
 * @param {*} test6 Test param
 * @param {"*"} test6 Test param
 * @param {?Number} test7 Test param
 * @param {...Number} test8 Test param
 * @param {!Number} test9 Test param
 * @param {String} test10 Test param
 * @param {Array} test11 Test param
 * @returns {Promise<Object<string, number|undefined>>} test return
 */`);

  expect(Result1).toMatchSnapshot();
});

test("Sould keep complex inner types", async () => {
  const Result1 = await subject(`/**
 * @param {Array<(String|Number)>} test test param
 * @param {Array<Object.<String, Number>>} test test param
 * @param {...Number} test Test param
 * @todo  todo is no param
 * @param {?Number} test Test param
 * @param {?undefined} test Test param
 * @param {!Number} test Test param
 * @param {Number} test Test param
 * @param {Number|String} test Test param
 * @param {undefined} test Test param
 * @param {*} test Test param
 */`);

  const Result2 = await subject(`/**
 * @returns {Promise<Object<string, number|undefined>>} test return
 */`);

  expect(Result1).toMatchSnapshot();
  expect(Result2).toMatchSnapshot();
});

test("Should align vertically param|property|returns|yields|throws if option set to true", async () => {
  const options = {
    jsdocVerticalAlignment: true,
  };
  const Result1 = await subject(
    `/**
 * @property {Object} unalginedProp unaligned property descriptin
 * @param {String} unalginedParam unaligned param description
 * @yields {Number} yields description
 * @returns {undefined}
 */`,
    options,
  );
  const Expected1 = `/**
 * @property {Object}    unalginedProp  Unaligned property descriptin
 * @param    {String}    unalginedParam Unaligned param description
 * @yields   {Number}                   Yields description
 * @returns  {undefined}
 */
`;

  const Result2 = await subject(
    `/**
 * @throws {CustomExceptio} unaligned throws description
 * @yields {Number} yields description
 * @return {String} unaligned returns description
 */`,
    options,
  );
  const Expected2 = `/**
 * @yields  {Number}         Yields description
 * @returns {String}         Unaligned returns description
 * @throws  {CustomExceptio} Unaligned throws description
 */
`;

  expect(Result1).toEqual(Expected1);
  expect(Result2).toEqual(Expected2);
});

test("Should align vertically param|property|returns|yields|throws if option set to true, and amount of spaces is different than default", async () => {
  const options1 = {
    jsdocVerticalAlignment: true,
    jsdocSpaces: 2,
  };
  const unformattedJsdoc = `/**
 * @property {Object} unalginedProp unaligned property descriptin
 * @param {String} unalginedParam unaligned param description
 * @throws {CustomExceptio} unaligned throws description
 * @yields {Number} yields description
 * @returns {undefined}
 */`;
  const Result1 = await subject(unformattedJsdoc, options1);

  const options2 = {
    jsdocVerticalAlignment: true,
    jsdocSpaces: 4,
  };
  const Result2 = await subject(
    `/**
 * @property {Object} unalginedProp unaligned property descriptin
 * @param {String} unalginedParam unaligned param description
 * @throws {CustomExceptio} unaligned throws description
 * @yields {Number} yields description
 * @returns {String} unaligned returns description
 */`,
    options2,
  );

  expect(Result1).toMatchSnapshot();
  expect(Result2).toMatchSnapshot();
});

test("Should insert proper amount of spaces based on option", async () => {
  const options1 = {
    jsdocSpaces: 2,
  };
  const Result1 = await subject(
    `/**
 * @param {Object} paramName param description that goes on and on and on utill it will need to be wrapped
 * @returns {Number} return description
 */`,
    options1,
  );

  const options2 = {
    jsdocSpaces: 3,
  };
  const Result2 = await subject(
    `/**
 * @param {Object} paramName param description that goes on and on and on utill it will need to be wrapped
 * @returns {Number} return description
 */`,
    options2,
  );

  expect(Result1).toMatchSnapshot();
  expect(Result2).toMatchSnapshot();
});

test("yields should work like returns tag", async () => {
  const options = {
    jsdocSpaces: 3,
  };
  const Result1 = await subject(
    `/**
 * @yields {Number} yields description
 */`,
    options,
  );

  const Result2 = await subject(
    `/**
 * @yield {Number} yields description
 */`,
    options,
  );

  const Result3 = await subject(
    `/**
 * @yield {Number}
 */`,
    options,
  );

  const Result4 = await subject(
    `/**
 * @yield yelds description
 */`,
    options,
  );

  const Result5 = await subject(
    `/**
 * @yield
 */`,
    options,
  );

  expect(Result1).toMatchSnapshot();
  expect(Result2).toMatchSnapshot();
  expect(Result3).toMatchSnapshot();
  expect(Result4).toMatchSnapshot();
  expect(Result5).toMatchSnapshot();
});
test("Big single word", async () => {
  const result = await subject(
    `/**
    * Simple Single Word
    * https://github.com/babel/babel/pull/7934/files#diff-a739835084910b0ee3ea649df5a4d223R67
   */`,
  );

  expect(result).toMatchSnapshot();
});

test("Hyphen at the start of description", async () => {
  const result = await subject(`
/**
 * Assign the project to an employee.
 * @param {Object} employee - The employee who is responsible for the project.
 * @param {string} employee.name - The name of the employee.
 * @param {string} employee.department - The employee's department.
 */
`);

  expect(result).toMatchSnapshot();
});

test("Bad defined name", async () => {
  const result = await subject(`
  /** @type{import('@jest/types/build/Config').InitialOptions} */
  /** @type{{foo:string}} */

  /** @typedef{import('@jest/types/build/Config').InitialOptions} name a description  */
`);

  expect(result).toMatchSnapshot();
});

test("Long description memory leak", async () => {
  const result = await subject(`
  /** Configures custom logging for the {@link @microsoft/signalr.HubConnection}.
   *
   * https://example.com
   * @param {LogLevel | string | ILogger} logging A {@link @microsoft/signalr.LogLevel}, a string representing a LogLevel, or an object implementing the {@link @microsoft/signalr.ILogger} interface.
   *    See {@link https://docs.microsoft.com/aspnet/core/signalr/configuration#configure-logging|the documentation for client logging configuration} for more details.
   * @returns The {@link @microsoft/signalr.HubConnectionBuilder} instance, for chaining.
   */
`);

  expect(result).toMatchSnapshot();
});

test("since ", async () => {
  const result = await subject(`
  /**
   * @since 3.16.0
   */
`);

  expect(result).toMatchSnapshot();
});

test("Incorrect comment", async () => {
  const result = await subject(`
  /***
   * Some comment
   */
  export class Dummy {}
  `);

  const result2 = await subject(`
  /**
   *
   */
  export class Dummy {}
  `);

  expect(result).toMatchSnapshot();
  expect(result2).toMatchSnapshot();
});

test("Empty comment", async () => {
  const result = await subject(`
  // Line Comment
  //
  `);

  expect(result).toMatchSnapshot();
});

test("Optional parameters", async () => {
  const result = await subject(`
  /**
   * @param {number=} arg1
   * @param {number} [arg2]
   * @param {number} [arg3=4]
   */
  `);

  expect(result).toMatchSnapshot();
});

test("Non-jsdoc comment", async () => {
  const result = await subject(`
  // @type   { something  }
  /* @type   { something  }  */
  /* /** @type   { something  }  */
  `);

  expect(result).toMatchSnapshot();
});

test("Format rest parameters properly", async () => {
  const result = await subject(`
  /**
   * @param {... *} arg1
   * @param {... number} arg2
   * @param {... (string|number)} arg3
   * @param {... string|number} arg4 This is equivalent to arg3
   *
   */
  function a(){}
  `);

  expect(result).toMatchSnapshot();
});

test("Line ends", async () => {
  const text = `
  /**
   * Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua.
   * @param {Object} paramName param description that goes on and on and on until it will need to be wrapped
   *
   */
  function a(){}`;
  const formatted = `/**
 * Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod
 * tempor incididunt ut labore et dolore magna aliqua.
 *
 * @param {Object} paramName Param description that goes on and on and on
 *   until it will need to be wrapped
 */
function a() {}
`;

  const text_crlf = text.replace(/\n/g, "\r\n");
  const text_lf = text;
  const formatted_crlf = formatted.replace(/\n/g, "\r\n");
  const formatted_lf = formatted;

  expect(await subject(text_lf, { endOfLine: "crlf" })).toEqual(formatted_crlf);
  expect(await subject(text_lf, { endOfLine: "lf" })).toEqual(formatted_lf);

  expect(await subject(text_crlf, { endOfLine: "crlf" })).toEqual(formatted_crlf);
  expect(await subject(text_crlf, { endOfLine: "lf" })).toEqual(formatted_lf);

  expect(await subject(text_lf, { endOfLine: "auto" })).toEqual(formatted_lf);
  expect(await subject(text_crlf, { endOfLine: "auto" })).toEqual(formatted_crlf);
});

test("param order", async () => {
  const result = await subject(`
  /**
* @param {  string   }    param0 description
* @param {  number   }    param2 description
* @param {  object   }    param1 description
   */
function fun(param0, param1, param2){}

export const SubDomain = {
/**
 * @param {} subDomainAddress2
 * @param {any} subDomainAddress
* @returns {import('axios').AxiosResponse<import('../types').SubDomain>} 
*/
async subDomain(subDomainAddress2,subDomainAddress) {
},
};
 

/**
 * @param {  string   }    param0 description
 * @param {  number   }    param2 description
 * @param {  object   }    param1 description
    */
 const fun=(param0, param1, param2)=>{
   console.log('')
 }
  
`);

  expect(result).toMatchSnapshot();

  const result2 = await subject(
    `
 /**
 * @param {object} c      - Options.
 * @param {string} a      - A for foo.
 * @param {object} opts   - Options.
 * @param {string} opts.b - Option b.
 * @param {string} opt2 - Option b.
 * @param {string} v - Option b.
 */
const foo = (a,c, { b }, v, opt2) => {
  console.log(a,v,c, b);
};
foo('a', { b: 'b' });
  
`,
    {
      jsdocVerticalAlignment: true,
    },
  );

  expect(result2).toMatchSnapshot();

  const result3 = await subject(
    `
 /**
 * @param {  string   }    param0 description
 * @param {  number   }    param2 description
 * @param {  object   }    param1 description
    */
 const fun=((param0: ()=>{}, param1:number, param2)=>{
   console.log('')
 })

/**
 * @param {  string   }    param0 description
 * @param {  number   }    param2 description
 * @param {  object   }    param1 description
    */
 function fun(param0:string, param1:{}, param2:()=>{}){}
 `, 
 {
   parser: 'babel-ts'
 })

 expect(result3).toMatchSnapshot()

});

test("jsdoc tags", async () => {
  const result2 = await subject(
    `/**
    * @namespace
    * @borrows trstr as trim
    */   

  /**
    * Whether the type should be non null, \`required: true\` = \`nullable: false\`
    *
    * @default (depends on whether nullability is configured in type or schema)
    * @see declarativeWrappingPlugin
    */
`,
  );

  expect(result2).toMatchSnapshot();

  const result3 = await subject(
    `
/**
 * @default 'i am a value' i am the description
*/   
`,
  );

  expect(result3).toMatchSnapshot();
});

test("example ", async () => {
  const result2 = await subject(
    `
/**
 * ABCCCC
 *
 * @example <caption>DDDD</caption>
 *
 *   const MyHook = () => {
 *     const config useConfig(state => state.config)
 *     return <span></span>
 *   }
 *
 *   const useMyHook = () => {
 *     const config useConfig(state => state.config)
 *     return config
 *   }
 *
 * @example <caption>AAAA</caption>
 *   const config = useConfig.getState().config
 */
`,
  );

  expect(await subject(await subject(result2))).toMatchSnapshot();
});

test("example with tab intention", async () => {
  const result2 = await subject(
    `
/**
 * @example
 * 	function Hello() {
 * 		console.log("Hello World");
 * 	}
 */
    
`,
    {
      useTabs: true,
    },
  );

  expect(
    await subject(
      await subject(result2, {
        useTabs: true,
      }),
      {
        useTabs: true,
      },
    ),
  ).toMatchSnapshot();
});

test("Optional params", async () => {
  const result = await subject(`
/**
 * @param {string=} p2 - An optional param (Google Closure syntax)
 * @param {string} [p3] - Another optional param (JSDoc syntax).
 * @returns {string=} 
*/
`);

  expect(result).toMatchSnapshot();
});

test("non-escapable character", async () => {
  const result = await subject(`
  /**
   * \\\\
   * 
   * 
   * \\\\-
   */
  `);

  expect(result).toMatchSnapshot();

  const result2 = await subject(`
  /**
   * \\\\
   */
  `);

  expect(result2).toMatchSnapshot();
});

test("@file", async () => {
  const result = await subject(`
  /** @file A file description */
`);

  expect(result).toMatchSnapshot();
});

test("Block quote", async () => {
  const result = await subject(`
  /** > A block quote */
`);

  expect(result).toMatchSnapshot();

  const result2 = await subject(`
  /**
   *  > \`\`\`js
   *  > > A block quote
   *  > \`\`\`
   *  > 
   *  > turns into
   *  > 
   *  > \`\`\`js
   *  > A block quote 
   *  > \`\`\`
   *  
   *  sdssasdassd
   *   
   */
`);

  expect(result2).toMatchSnapshot();
});

test("File with just an import", async () => {
  const result = await subject(
    `
import { something } from './index';
`,
    {
      jsdocDescriptionWithDot: true,
      jsdocSingleLineComment: false,
      jsdocSeparateTagGroups: true,
      jsdocPreferCodeFences: true,
      tsdoc: true,
    },
  );

  expect(await subject(await subject(result))).toMatchSnapshot();
});
