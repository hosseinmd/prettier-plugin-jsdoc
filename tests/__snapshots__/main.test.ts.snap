// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[` undefined|null|void type 1`] = `
"/** @returns {undefined} */
"
`;

exports[` undefined|null|void type 2`] = `
"/** @returns {null} */
"
`;

exports[` undefined|null|void type 3`] = `
"/** @returns {void} */
"
`;

exports[`Bad defined name 1`] = `
"/** @type {import(\\"@jest/types/build/Config\\").InitialOptions} */
/** @type {{ foo: string }} */

/** @typedef {import(\\"@jest/types/build/Config\\").InitialOptions} name A description */
"
`;

exports[`Big single word 1`] = `
"/**
 * Simple Single Word
 * https://github.com/babel/babel/pull/7934/files#diff-a739835084910b0ee3ea649df5a4d223R67
 */
"
`;

exports[`Empty comment 1`] = `
"// Line Comment
//
"
`;

exports[`Format rest parameters properly 1`] = `
"/**
 * @param {...any} arg1
 * @param {...number} arg2
 * @param {...(string | number)} arg3
 * @param {...(string | number)} arg4 This is equivalent to arg3
 */
function a() {}
"
`;

exports[`Hyphen at the start of description 1`] = `
"/**
 * Assign the project to an employee.
 *
 * @param {Object} employee - The employee who is responsible for the project.
 * @param {string} employee.name - The name of the employee.
 * @param {string} employee.department - The employee's department.
 */
"
`;

exports[`Incorrect comment 1`] = `
"/** Some comment */
export class Dummy {}
"
`;

exports[`Incorrect comment 2`] = `
"export class Dummy {}
"
`;

exports[`JS code should be formatted as usuall 1`] = `
"const variable1 = 1; // No semicolon
const stringVar = \\"text\\"; // Wrong quotes
const indented = 2; // Wrong indentation

// Longer then 80 characters
const someLongList = [
  \\"private\\",
  \\"memberof\\",
  \\"description\\",
  \\"example\\",
  \\"param\\",
  \\"returns\\",
  \\"link\\",
];
"
`;

exports[`Long description memory leak 1`] = `
"/**
 * Configures custom logging for the {@link @microsoft/signalr.HubConnection}.
 *
 * https://example.com
 *
 * @param {LogLevel | string | ILogger} logging A
 *   {@link @microsoft/signalr.LogLevel}, a string representing a LogLevel, or
 *   an object implementing the {@link @microsoft/signalr.ILogger} interface.
 *   See
 *   {@link https://docs.microsoft.com/aspnet/core/signalr/configuration#configure-logging|the documentation for client logging configuration}
 *   for more details.
 * @returns The {@link @microsoft/signalr.HubConnectionBuilder} instance, for chaining.
 */
"
`;

exports[`Non-jsdoc comment 1`] = `
"// @type   { something  }
/* @type   { something  }  */
/* /** @type   { something  }  */
"
`;

exports[`Optional parameters 1`] = `
"/**
 * @param {number} [arg1]
 * @param {number} [arg2]
 * @param {number} [arg3=4] Default is \`4\`
 */
"
`;

exports[`Should align vertically param|property|returns|yields|throws if option set to true, and amount of spaces is different than default 1`] = `
"/**
 * @property  {Object}          unalginedProp   Unaligned property descriptin
 * @param     {String}          unalginedParam  Unaligned param description
 * @yields    {Number}                          Yields description
 * @returns   {undefined}
 * @throws    {CustomExceptio}                  Unaligned throws description
 */
"
`;

exports[`Should align vertically param|property|returns|yields|throws if option set to true, and amount of spaces is different than default 2`] = `
"/**
 * @property    {Object}            unalginedProp     Unaligned property descriptin
 * @param       {String}            unalginedParam    Unaligned param description
 * @yields      {Number}                              Yields description
 * @returns     {String}                              Unaligned returns description
 * @throws      {CustomExceptio}                      Unaligned throws description
 */
"
`;

exports[`Should convert to single line if necessary 1`] = `
"/** Single line description */
"
`;

exports[`Should convert to single line if necessary 2`] = `
"/** Single line description */
"
`;

exports[`Should convert to single line if necessary 3`] = `
"/**
 * Single line description
 *
 * @returns {Boolean} Always true
 */
"
`;

exports[`Should convert to single multiLine 1`] = `
"/**
 * Single line description
 */
"
`;

exports[`Should convert to single multiLine 2`] = `
"/**
 * Single line description
 */
"
`;

exports[`Should convert to single multiLine 3`] = `
"/**
 * Single line description
 *
 * @returns {Boolean} Always true
 */
"
`;

exports[`Should format jsDoc default values 1`] = `
"/**
 * @param {String} [arg1=\\"defaultTest\\"] Foo. Default is \`\\"defaultTest\\"\`
 * @param {number} [arg2=123] The width of the rectangle. Default is \`123\`
 * @param {number} [arg3=123] Default is \`123\`
 * @param {number} [arg4=Foo.bar.baz] Default is \`Foo.bar.baz\`
 * @param {number | string} [arg5=123] Something. Default is \`123\`
 */
"
`;

exports[`Should format jsDoc default values 2`] = `
"/**
 * @param {String} [arg1=\\"defaultTest\\"] Foo. Default is \`\\"defaultTest\\"\`
 * @param {number} [arg2=123] The width of the rectangle. Default is \`123\`
 * @param {number} [arg3=123] Default is \`123\`
 * @param {number} [arg4=Foo.bar.baz] Default is \`Foo.bar.baz\`
 * @param {number | string} [arg5=123] Something. Default is \`123\`
 */
"
`;

exports[`Should format regular jsDoc 1`] = `
"/**
 * Function example description that was wrapped by hand so it have more then
 * one line and don't end with a dot REPEATED TWO TIMES BECAUSE IT WAS EASIER to
 * copy function example description that was wrapped by hand so it have more
 * then one line.
 *
 * @async
 * @private
 * @memberof test
 * @example
 *   var one = 5;
 *   var two = 10;
 *
 *   if (one > 2) {
 *     two += one;
 *   }
 *
 * @param {String | Number} text - Some text description that is very long and
 *   needs to be wrapped
 * @param {String} [defaultValue=\\"defaultTest\\"] TODO. Default is \`\\"defaultTest\\"\`
 * @param {Number | Null} [optionalNumber]
 * @returns {Boolean} Description for @returns with s
 * @undefiendTag
 * @undefiendTag {number} name des
 */
const testFunction = (text, defaultValue, optionalNumber) => true;
"
`;

exports[`Should format regular jsDoc 2`] = `
"/**
 * Function example description that was wrapped by hand so it have more then
 * one line and don't end with a dot REPEATED TWO TIMES BECAUSE IT WAS EASIER to
 * copy function example description that was wrapped by hand so it have more
 * then one line.
 *
 * @async
 * @private
 * @memberof test
 * @example
 *   var one = 5;
 *   var two = 10;
 *
 *   if (one > 2) {
 *     two += one;
 *   }
 *
 * @param {String | Number} text - Some text description that is very long and
 *   needs to be wrapped
 * @param {String} [defaultValue=\\"defaultTest\\"] TODO. Default is \`\\"defaultTest\\"\`
 * @param {Number | Null} [optionalNumber]
 * @returns {Boolean} Description for @returns with s
 * @undefiendTag
 * @undefiendTag {number} name des
 */
const testFunction = (text, defaultValue, optionalNumber) => true;
"
`;

exports[`Should insert proper amount of spaces based on option 1`] = `
"/**
 * @param  {Object}  paramName  Param description that goes on and on and on
 *   utill it will need to be wrapped
 * @returns  {Number}  Return description
 */
"
`;

exports[`Should insert proper amount of spaces based on option 2`] = `
"/**
 * @param   {Object}   paramName   Param description that goes on and on and on
 *   utill it will need to be wrapped
 * @returns   {Number}   Return description
 */
"
`;

exports[`Should keep defined inner types 1`] = `
"/** @param {String[]} test Test param */
"
`;

exports[`Should keep defined inner types 2`] = `
"/** @param {String[]} test Test param */
"
`;

exports[`Should keep defined inner types 3`] = `
"/** @param {(String | Object)[]} test Test param */
"
`;

exports[`Should keep defined inner types 4`] = `
"/** @returns {Promise<Number | String | undefined>} Test promise */
"
`;

exports[`Should keep defined inner types 5`] = `
"/** @returns {Object<Number | String | undefined>} Test object */
"
`;

exports[`Sould keep complex inner types 1`] = `
"/**
 * @param {(String | Number)[]} test Test param
 * @param {Object<String, Number>[]} test Test param
 * @param {...Number} test Test param
 * @param {Number | null} test Test param
 * @param {undefined | null} test Test param
 * @param {!Number} test Test param
 * @param {Number} test Test param
 * @param {Number | String} test Test param
 * @param {undefined} test Test param
 * @param {any} test Test param
 * @todo Todo is no param
 */
"
`;

exports[`Sould keep complex inner types 2`] = `
"/** @returns {Promise<Object<string, number | undefined>>} Test return */
"
`;

exports[`Sould keep params ordering when more than 10 tags are present 1`] = `
"/**
 * Description
 *
 * @param {Number} test1 Test param
 * @param {Number} test2 Test param
 * @param {Number | String} test3 Test param
 * @param {undefined | null} test4 Test param
 * @param {!undefined} test5 Test param
 * @param {any} test6 Test param
 * @param {\\"*\\"} test6 Test param
 * @param {Number | null} test7 Test param
 * @param {...Number} test8 Test param
 * @param {!Number} test9 Test param
 * @param {String} test10 Test param
 * @param {Array} test11 Test param
 * @returns {Promise<Object<string, number | undefined>>} Test return
 */
"
`;

exports[`example  1`] = `
"/**
 * ABCCCC
 *
 * @example <caption>DDDD</caption>
 *   const MyHook = () => {
 *   const config useConfig(state => state.config)
 *   return <span></span>
 *   }
 *
 *   const useMyHook = () => {
 *   const config useConfig(state => state.config)
 *   return config
 *   }
 *
 * @example <caption>AAAA</caption>
 *   const config = useConfig.getState().config;
 */
"
`;

exports[`example with tab intention 1`] = `
"/**
 * @example
 * 	function Hello() {
 * 		console.log(\\"Hello World\\");
 * 	}
 */
"
`;

exports[`jsdoc tags 1`] = `
"/**
 * @namespace
 * @borrows trstr as trim
 */

/**
 * Whether the type should be non null, \`required: true\` = \`nullable: false\`
 *
 * @default (depends on whether nullability is configured in type or schema)
 * @see declarativeWrappingPlugin
 */
"
`;

exports[`param order 1`] = `
"/**
 * @param {string} param0 Description
 * @param {object} param1 Description
 * @param {number} param2 Description
 */
function fun(param0, param1, param2) {}

export const SubDomain = {
  /**
   * @param subDomainAddress2
   * @param {any} subDomainAddress
   * @returns {import(\\"axios\\").AxiosResponse<import(\\"../types\\").SubDomain>}
   */
  async subDomain(subDomainAddress2, subDomainAddress) {},
};

/**
 * @param {string} param0 Description
 * @param {object} param1 Description
 * @param {number} param2 Description
 */
function fun(param0: string, param1: {}, param2: () => {}) {}

/**
 * @param {string} param0 Description
 * @param {object} param1 Description
 * @param {number} param2 Description
 */
const fun = (param0, param1, param2) => {
  console.log(\\"\\");
};

/**
 * @param {string} param0 Description
 * @param {number} param2 Description
 * @param {object} param1 Description
 */
const fun = (param0: () => {}, param1: number, param2) => {
  console.log(\\"\\");
};
"
`;

exports[`param order 2`] = `
"/**
 * @param {string} a      - A for foo.
 * @param {object} c      - Options.
 * @param {object} opts   - Options.
 * @param {string} opts.b - Option b.
 * @param {string} v      - Option b.
 * @param {string} opt2   - Option b.
 */
const foo = (a, c, { b }, v, opt2) => {
  console.log(a, v, c, b);
};
foo(\\"a\\", { b: \\"b\\" });
"
`;

exports[`since  1`] = `
"/** @since 3.16.0 */
"
`;

exports[`yields should work like returns tag 1`] = `
"/** @yields   {Number}   Yields description */
"
`;

exports[`yields should work like returns tag 2`] = `
"/** @yields   {Number}   Yields description */
"
`;

exports[`yields should work like returns tag 3`] = `
"/** @yields   {Number} */
"
`;

exports[`yields should work like returns tag 4`] = `
"/** @yields   Yelds description */
"
`;

exports[`yields should work like returns tag 5`] = `
"/** @yields */
"
`;
